Android lib link statically/dynamically

arthur740212 opened the pull request:
Toggle for linking eos library statically in config editor
The target eos-sdk would be added from PlatformSpecificStreamingAssets to Plugin/Android folder
according to the selected method

arthur740212 committed:
feat(android,lib) : library static/dynamic linking options

This allows user to select to use dynamic or static linking for android eos-sdk library

arthur740212 committed:
assets(android,lib) : eos-sdk libs (static and dynamic))

andrew-hirata-playeveryware commented:
The assets that were added to this review need to be moved from the ```Assets``` directory and into ```PlatformSpecificAssets/EOS/Android```, which is one directory up, otherwise I believe both will be included in the build.

andrew-hirata-playeveryware commented on Assets/Plugins/AndroidEditor/Editor/EOSOnPreprocessBuild_android.cs:
```diff
@@ -221,4 +223,32 @@ public void ConfigureEOSDependentLibrary()
+        EOSConfigFile<EOSAndroidConfig> configFile = new EOSConfigFile<EOSAndroidConfig>(configFilenamePath);
+        configFile.LoadConfigFromDisk();
+
+        var sdkAndroidPath = System.IO.Path.Combine("Assets", "PlatformSpecificStreamingAssets", "EOS", "Android");
```
The path here will need to be corrected when the added assets are moved to the correct location.

andrew-hirata-playeveryware replied:
Additionally, this needs to be modified such that it works when the code is run from the plugin directory, which has a different path layout. The general way to get files relative the package install location is by doing something like this: 

```
string packagePathname = Path.GetFullPath("Packages/" + GetPackageName() + "/PlatformSpecificStreamingAssets~/" + subpath);
```

Where `GetPackageName` is defined like so:
```
    private static string GetPackageName()
    {
        return "com.playeveryware.eos";
    }
```

While not strictly required, you can also use the following to find the assets, so that it's not 'hard coded' to the asset's directory:
```
string streamingAssetsSamplesPathname = Path.Combine(Application.dataPath, "../PlatformSpecificStreamingAssets/" + subpath);
```
As you'll see here, the second parameter into the `Path.Combine` method has a '..' in front of it, which allows one to find sibling directories to the `Assets` directory.

andrew-hirata-playeveryware commented on Assets/Plugins/AndroidEditor/Editor/PlatformSpecificConfigEditorAndroid.cs:
```diff
@@ -69,6 +69,8 @@ public void OnGUI()
                 overrideValues = configFile.currentEOSConfig.overrideValues;
             }
 
+            EpicOnlineServicesConfigEditor.AssigningBoolField("Link EOS-SDK Statically", ref configFile.currentEOSConfig.linkLibStatically);
```
This config option should be put in a plugin editor config section, similar to how `EOSPluginEditorToolsConfig` is done, as this config option is ideally for runtime only specific configuration.

andrew-hirata-playeveryware commented on Assets/Plugins/Android/EOSAndroidConfig.cs:
```diff
@@ -34,6 +34,7 @@ public class EOSAndroidConfig : ICloneableGeneric<EOSAndroidConfig>, IEmpty
 
         public EOSConfig overrideValues;
 
+        public bool linkLibStatically = false;
```
As mentioned in the PlatformSpecificConfigEditorAndroid comment, this should preferably be in it's own IEOSPluginEditorConfigurationSection for android options

arthur740212 committed:
fix(ConfigEditor,Android) : Move static/dynamic link library settings to plugin editor config section

Move static/dynamic link library settings to plugin editor config section as this config option is ideally for runtime only specific configuration

arthur740212 committed:
fix(android,config editor) : Revert settings in platformSpecific for android lib linking

arthur740212 committed:
assets(EOSlib,android) : Move them to PlatformSpecificAssets

arthur740212 committed:
fix(android) : filepath and config settings modification for library linking

Code modification with respect to Android EOS sdk filepath and config editor changes

andrew-hirata-playeveryware commented on Assets/Plugins/AndroidEditor/Editor/EOSOnPreprocessBuild_android.cs:
```diff
@@ -221,4 +223,34 @@ public void ConfigureEOSDependentLibrary()
+        var libSubpath = "/aar/eos-sdk.aar";
+
+        //string packagePathname = Path.GetFullPath("Packages/" + GetPackageName() + "/PlatformSpecificAssets~/" );
+        string dynamicPathName = Path.Combine(Application.dataPath, androidAssetFilepath, "dynamic-stdc++" + libSubpath);
```
This still needs to have changes done so that it handles the case where it's running in the plugin, as the path names aren't shared. One can get the plugin path in the way described here https://github.com/PlayEveryWare/eos_plugin_for_unity/pull/178#discussion_r1046513278

arthur740212 committed:
fix : Copy library through package

andrew-hirata-playeveryware commented on Assets/Plugins/Android/Editor/EOSOnPreprocessBuild_android.cs:
```diff
@@ -221,4 +223,33 @@ public void ConfigureEOSDependentLibrary()
+        string libSubpath = "/aar/eos-sdk.aar";
+
+        string dynamicPathName = Path.Combine(Application.dataPath, packagePathname, "dynamic-stdc++" + libSubpath);
+        string staticPathName = Path.Combine(Application.dataPath, packagePathname, "static-stdc++" + libSubpath);
```
This is checking for the paths incorrectly.
It needs check if the libraries exist in the `packagePathname` directory, and if they don't, assume they exist in the `../PlatformSpecificAssets/EOS/Android` path and copy them into `destPathname`.

arthur740212 committed:
fix(android) : detects if library is copied from package

arthur740212 committed:
fix(syntax) : parameter order swap

arthur740212 committed:
fix(package description) : Remove path sources that don't exist anymore

arthur740212 committed:
Merge branch 'development' into android_lib_link_static_or_dynamic

arthur740212 committed:
Revert "Merge branch 'development' into android_lib_link_static_or_dynamic"

This reverts commit 23da0d8712cfa5e9d5512ecac9af1122dde220d5, reversing
changes made to f729699b63fd1d3ece433b8786e330cb29132f02.

arthur740212 committed:
Revert "Revert "Merge branch 'development' into android_lib_link_static_or_dynamic""

This reverts commit 219e7979d145bf36dcb1a093fb32f73304ac9e98.

arthur740212 committed:
fix : Merge Conflict by removed functions

andrew-hirata-playeveryware reviewed (approved):
I think this needs to have the conflict resolved and then be squashed, but this code otherwise looks good to merge.

arthur740212 committed:
Update EOSOnPreprocessBuild_android.cs

arthur740212 committed:
Merge branch 'development' into android_lib_link_static_or_dynamic
